### DISCUSSION TITLE
Add (Micro)charts to the .NET MAUI Community Toolkit

### DISCUSSION THREAD

**Discussion by jfversluis** [CommunityToolkit/Maui#232] [New Feature Discussions]
Over the weekend there was [this tweet](https://twitter.com/ACaliaro/status/1477545571120562177?s=20) mentioning that the library [Microcharts](https://github.com/microcharts-dotnet/Microcharts) is looking for a new maintainer. From the Microcharts repo:

<p align="center">
<img src="https://github.com/microcharts-dotnet/Microcharts/raw/main/assets/animations.gif"  />
</p>

> Microcharts is an extremely simple charting library for a wide range of platforms (see Compatibility section below), with shared code and rendering for all of them!

The compatibility section that is referenced:

> Built in views are provided for UWP, Xamarin.Forms, Xamarin.iOS and Xamarin.Android, Xamarin.macOS, but any other .NET Standard 2.0 SkiaSharp supported platform is also compatible (...)

This triggered me into thinking: maybe this would be a nice addition to the .NET MAUI Community Toolkit or maybe the bigger Toolkit community so other platforms can leverage it as well, since it's based on [SkiaSharp](https://github.com/mono/SkiaSharp), that should be possible. So opening the discussion here to hear thoughts. A couple of mine:

* If we were to do this (if!) then I think we should consider looking into basing it on [Microsoft.Maui.Graphics](https://github.com/dotnet/Microsoft.Maui.Graphics)
* Additionally, because of dependencies we're bringing in, I guess we should make this a separate package to not burden people that have no interest in charts with extra dependencies
* Investigate if we can make this broader than just .NET MAUI so also Windows can benefit and maybe others in the future
* Ideally we should get in touch with Ed (current maintainer) and see if we can reuse whatever he has to not reinvent everything
* We will only implement this for .NET MAUI (or broader), not Xamarin.Forms

What are peoples thoughts on this?

**Comment by krdmllr** [comment]
While making it available on other platforms sounds great, I would prefer to avoid using skia due to the libraries size. ~~I just checked and saw that Maui Graphics is using skia as well so the only option to avoid skia would be to use the platforms native drawing APIs, which sounds like a lot of work.~~ 

~~In that case, just keeping skia would probably be the best option. Since skia supports all platforms, adding microcharts to the whole community toolkit like Gerald suggested + having an abstraction for Maui views might be the way to go?~~

I missread the docs, so +1 to use Microsoft.Maui.Graphics.

**Reply by mgierlasinski** [reply]
Microsoft.Maui.Graphics is using native drawing, it doesn't have dependency on SkiaSharp. You can optionally use package Microsoft.Maui.Graphics.Skia to have Skia rendering backend.

Rewriting Microcharts to draw with Maui.Graphics seems to be the most flexible way as you rely only on ICanvas abstraction.

**Reply by saint4eva** [reply]
> Maui Graphics is using skia as well

Microsoft.Maui.Graphics does not have any dependency on SkiaSharp. Kindly take a look at the repo.

**Reply by krdmllr** [reply]
Ah, I got confused by the supported platforms section in the readme. Then Maui.Graphics would obviously be a great choice.

**Comment by saint4eva** [comment]
> What are peoples thoughts on this?

- Implement it for .NET Maui and not for Xamarin.Form
- Use Microsoft.Maui.Graphics
- Make it a separate package so other frameworks and runtimes could use it.

**Comment by acaliaro** [comment]
There is also this

https://github.com/jsuarezruiz/TemplateUI#datavisualization

**Comment by lietu** [comment]
Just my quick 2 cents as a complete outsider:

I was just evaluating MAUI vs Flutter for a small project, everything was leaning towards it making sense to try out MAUI even in the early preview stages, but then I did one last check - can I find an easy way to make decent enough charts.

For MAUI I found 3, maybe 4? separate commercial solutions, all say "free trial" and none list pricing, i.e. they're out of my budget range.

For Flutter I found the free `charts_flutter` library in the Dart package directory as the first search result, free, capable, supports all target platforms.

While there would be other benefits to MAUI/C#, I think this is what ends up making the decision for me.

**Reply by jfversluis** [reply]
Thanks for your input! Very valuable! :)

I think there is one nuance here though, Flutter has been out for a while and .NET MAUI is still in preview so it kind of makes sense that there might not be a ton of libraries for it now. But if you have to start building now, then sure, I understand that you would go with Flutter if this is important to you.

This does underline the importance of having something like this for .NET MAUI asap though. Thanks!

**Comment by giuseppenovielli** [comment]
https://blog.dangl.me/archive/building-hacky-but-free-graphs-and-charts-in-xamarinforms-with-chartjs/

For a better charts, with a a lot of functions, you can use Xamarin.Forms/.NET MAUI WebView, with

https://www.chartjs.org/

or even better -> https://d3js.org/

With no dependency from Skia or other, only webview

Yes, gesture into webview aren't the best, but these library are very complete and customizable!
And WebView is ready for all platform!

This can be also an alternative that i want to show you!

**Comment by GreenShadeZhang** [comment]
I think it's a good idea to use Microsoft.Maui.Graphics

**Comment by JORGEGO** [comment]
Let me ask about decision taken around the Charts in MAUI 
What it is the recommendation for organization that have used Microcharts with Xamarin.Forms and now have plans to migrate to MAUI, and must be free controls.
Thanks

**Reply by michael-hawker** [reply]
Ideally someone can help the Microcharts folks finish the MAUI PR: https://github.com/microcharts-dotnet/Microcharts/pull/280, then you can just migrate your codebase as-is.

**Reply by bijington** [reply]
Great point! I don't have the bandwidth right now but I would be keen to see if this could be moved forwards so will try and find some time.

**Reply by michael-hawker** [reply]
@bijington looks like they're just looking for folks to try out the PR here, so getting close: https://github.com/microcharts-dotnet/Microcharts/pull/325

**Reply by bijington** [reply]
> @bijington looks like they're just looking for folks to try out the PR here, so getting close: https://github.com/microcharts-dotnet/Microcharts/pull/325

Oh if it's just testing that's not so bad. I saw the 74 files changed and thought I don't know enough just yet and would need to look into the micro charts codebase. Testing I can do

**Comment by eman1986** [comment]
Maintainer of Microcharts here 🙋🏻‍♂️ 

I'd prefer those looking to keep Microcharts alive contribute to the repo, I don't like the idea of telling everyone using Xamarin.Forms, Xamarin.Android, Xamarin.iOS, and so forth that we're leaving you out because we're focused on MAUI now.

To me, if you want a maui only solution, then someone should make a maui only chart library.

The goal of Microcharts has always been that anything that works with Skia is welcomed on this library.

We had someone from the community add maui support, I essentially just need to look at another PR for it and then push up a preview release so the community can try it out and ensure it works and then I'll cut an official 1.0 release.

I am happy to see people are interested in seeing this library continue, I certainly don't want to see it die, I'm the 4th person to take over in its history as people left for various reasons, while I'm no longer working with .NET tech anymore, I do want to see this continue and happy to be a support role to ensure it survives.

**Reply by RonalPortes** [reply]
Has it been any progress in making this library available for MAUI?

**Reply by michael-hawker** [reply]
@RonalPortes see the Microcharts repo, the code in the repo supports MAUI, it just hasn't been published to NuGet, see this PR: https://github.com/microcharts-dotnet/Microcharts/pull/328

**Reply by RonalPortes** [reply]
> @RonalPortes see the Microcharts repo, the code in the repo supports MAUI, it just hasn't been published to NuGet, see this PR: https://github.com/microcharts-dotnet/Microcharts/pull/328

Thank you Michael.

**Reply by tibuprophen** [reply]
Just saw microcharts.Maui has been released 2 weeks ago 🎉  
See https://www.nuget.org/packages/Microcharts.Maui/

**Comment by jfversluis** [comment]
OK looks like there is movement in supporting MicroCharts for .NET MAUI and we're not necessarily looking for more work at the moment. Closing this for now, who knows what might happen in the future!

**Comment by AjitKRuin** [comment]
I have updated from Xamarin.IOS to .Net iOS , but I am not able to use MicroCharts.iOS namespace and ChartView. Is MichroCharts not available in .net8.0-iOS?

**Reply by jfversluis** [reply]
MicroCharts have never been made part of the .NET MAUI Community Toolkit. If you want to use it or wonder about the support, please check out the repository: https://github.com/microcharts-dotnet/Microcharts
